commands/deals.go

```diff
@@ -0,0 +1,108 @@
+package commands
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-ipfs-cmdkit"
+	"github.com/ipfs/go-ipfs-cmds"
+
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/porcelain"
+	"github.com/filecoin-project/go-filecoin/protocol/storage/storagedeal"
+)
+
+const (
+	clientOnly = "client"
+	minerOnly  = "miner"
+)
+
+var dealsCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline: "Manage and inspect deals made by or with this node",
+	},
+	Subcommands: map[string]*cmds.Command{
+		"list": dealsListCmd,
+	},
+}
+
+// DealsListResult represents the subset of deal data returned by deals list
+type DealsListResult struct {
+	Miner       address.Address   `json:"minerAddress"`
+	PieceCid    cid.Cid           `json:"pieceCid"`
+	ProposalCid cid.Cid           `json:"proposalCid"`
+	State       storagedeal.State `json:"state"`
+}
+
+var dealsListCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline: "List all deals",
+		ShortDescription: `
+Lists all recorded deals made by or with this node. This may include pending
+deals, active deals, finished deals and cancelled deals.
+`,
+	},
+	Options: []cmdkit.Option{
+		cmdkit.BoolOption(clientOnly, "c", "only return deals made as a client"),
+		cmdkit.BoolOption(minerOnly, "m", "only return deals made as a miner"),
+	},
+	Run: func(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment) error {
+		minerAddress, _ := GetPorcelainAPI(env).ConfigGet("mining.minerAddress")
+		dealsCh, err := GetPorcelainAPI(env).DealsLs(req.Context)
+		if err != nil {
+			return err
+		}
+
+		if filterForMiner, _ := req.Options[minerOnly].(bool); filterForMiner {
+			dealsCh = filterDealChannel(dealsCh, func(deal *storagedeal.Deal) bool {
```

Comment:
Creating a goroutine and channel here is over-complicated. The `filterDealChannel` function can be replaced with an if statement in the dealsCh loop using the filter function defined here or the one defined for the client.